File: LedgerKotlin/src/main/kotlin/com/ledger/app/services/ledger/LedgerService.kt
package com.ledger.app.services.ledger

import com.ledger.app.models.ledger.Entry
import com.ledger.app.models.ledger.Ledger
import com.ledger.app.models.ledger.Page
import com.ledger.app.models.ledger.Receipt

interface LedgerService {
    fun createLedger(name: String, linesPerPage: Int, hashAlgorithm: String)

    fun createEntry(ledgerName: String, content: String, senders: List<String>, recipients: List<String>, relatedEntries: List<String> = emptyList(), keywords: List<String> = emptyList()): Entry?
    fun signEntry(entryId: String, signerId: String, signature: String, publicKey: String, signingAlgorithm: String)
    fun eraseEntry(ledgerName: String, entryId: String, userId: String)
    fun restoreEntry(ledgerName: String, entryId: String, userId: String, originalContent: String)

    fun logSystemEvent(ledgerName: String, declaringService: String, userId: String?, details: String)

    fun addKeywords(entryId: String, keywords: List<String>)
    fun removeKeyword(entryId: String, keyword: String)
    fun addRelatedEntries(entryId: String, relatedEntries: List<String>)
    fun removeRelatedEntry(entryId: String, relatedEntry: String)

    // GETs

    fun getAvailableLedgers(): List<String>
    fun getLedger(ledgerName: String): Ledger?

    fun getPage(ledgerName: String, number: Int): Page?

    fun getEntry(entryId: String): Entry?
    fun getEntriesBySender(ledgerName: String, userId: String): List<Entry>
    fun getEntriesByRecipient(ledgerName: String, userId: String): List<Entry>
    fun getEntriesByKeyword(ledgerName: String, keyword: String): List<Entry>
    fun getRelatedEntries(entryId: String): List<Entry>
    fun getEntriesNeedingSignature(ledgerName: String, userId: String): List<Entry>

    fun getReceipt(userId: String, entryId: String): Receipt
}

File: Cliente/ledger_client/src/services/api/LedgerService.ts
import { Fetcher, ApiResponse } from './Fetcher';
import { LedgerDTO, PageDTO, PageSummaryDTO, EntryDTO } from '@/models/ledger_dto';

export class LedgerService {
  private baseUrl = "/ledger";
  private fetcher = new Fetcher();

  async getAvailableLedgers(): Promise<ApiResponse<string[]>> {
    return await this.fetcher.request<string[]>(`${this.baseUrl}/available`, {
      method: 'GET',
      requireAuth: true,
    });
  }

  async getLedger(ledgerName: string): Promise<ApiResponse<LedgerDTO>> {
    return await this.fetcher.request<LedgerDTO>(`${this.baseUrl}/${ledgerName}`, {
      method: 'GET',
      requireAuth: true,
    });
  }

  async getPage(ledgerName: string, pageNumber: number): Promise<ApiResponse<PageDTO>> {
    return await this.fetcher.request<PageDTO>(`${this.baseUrl}/${ledgerName}/page/${pageNumber}`, {
      method: 'GET',
      requireAuth: true,
    });
  }

  async getEntry(entryId: string): Promise<ApiResponse<EntryDTO>> {
    return await this.fetcher.request<EntryDTO>(`${this.baseUrl}/entry/${entryId}`, {
      method: 'GET',
      requireAuth: true,
    });
  }

  async addKeywords(entryId: string, keywords: string[]): Promise<ApiResponse<{ message: string }>> {
    return await this.fetcher.request<{ message: string }>(`${this.baseUrl}/entry/${entryId}/keywords`, {
      method: 'POST',
      body: keywords,
      requireAuth: true,
    });
  }

  async removeKeyword(entryId: string, keyword: string): Promise<ApiResponse<{ message: string }>> {
    return await this.fetcher.request<{ message: string }>(`${this.baseUrl}/entry/${entryId}/keyword/${keyword}`, {
      method: 'DELETE',
      requireAuth: true,
    });
  }
}